### PR TITLE
Refactor GetLocalCurrency

### PR DESCRIPTION
**Pull request by luacmartins**
### Details
- Creates OpenApp and ReconnectApp
- Removes usages of `fetchLocalCurrency` from App init
- Removes usages of `getCurrencyList` since we only want to fetch it on OpenApp

Holding on https://github.com/Expensify/App/pull/9560 and https://github.com/Expensify/Web-Expensify/pull/34151 hitting prod

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/213883

### Tests
1. Pull the changes from https://github.com/Expensify/Web-Expensify/pull/34151
1. Open the Network tab in the browser console.
2. Clear cache and browsing history for the App.
3. Sing in
4. Verify that OpenApp is fired and that the `localCurrencyCode` and `currencyList` are returned
5. Verify that those values are set in the `personalDetails` and `currencyList` keys
6. Navigate to `+ > Request money`
7. Verify that your local currency shows up
8. Click on the currency symbol and verify that you can see the currency list.
9. Set the App to offline and back online in the browser console.
10. Verify that the `ReconnectApp` API request was triggered.

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
1. Clear cache and browsing history for the App.
2. Sing in
3. Navigate to `+ > Request money`
4. Verify that your local currency shows up
5. Click on the currency symbol and verify that you can see the currency list.
9. Set the App to offline and back online in the browser console.
10. Repeat steps 3-5

- [ ] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/22219519/176318578-17ccd5d2-0255-43a6-9180-c02cbd6ae5fb.mov

#### Mobile Web

https://user-images.githubusercontent.com/22219519/176320083-5c974579-77f8-4e3d-9390-7d6c5e22b73b.mov

#### Desktop

https://user-images.githubusercontent.com/22219519/176320103-84e7ec00-cd78-4f90-ab63-48970da77d6f.mov

#### iOS

https://user-images.githubusercontent.com/22219519/176320137-64af56ae-34b3-4fb1-8ff8-3398b660de97.mov

#### Android

https://user-images.githubusercontent.com/22219519/176320471-dd0c733e-4fb5-48fd-9a55-0327ec449854.mov

### CODE REVIEW

**Comment by luacmartins** [comment]
I'd wait since this PR might change a bit based on that comment.

File: src/libs/actions/PersonalDetails.js [hunk 1]
```diff
@@ -323,7 +310,6 @@ function fetchLocalCurrency() {
         .then((data) => {
             currency = data.currency;
         })
-        .then(getCurrencyList)
         .then(() => {
             Onyx.merge(ONYXKEYS.MY_PERSONAL_DETAILS, {localCurrencyCode: currency});
```
Comment (mountiny) [inline src/libs/actions/PersonalDetails.js:314]: @luacmartins What I mentioned [here](https://github.com/Expensify/Web-Expensify/pull/34151#discussion_r919249607), this is how we currently store and merge the data about `localCurrencyCode`

Comment (luacmartins) [reply src/libs/actions/PersonalDetails.js:314]: Yea, I'm planning on changing this PR based on the changes [here](https://github.com/Expensify/App/pull/9560)

**Comment by luacmartins** [comment]
@mountiny @PauloGasparSv this is ready for review! We are still holding on Web-E hitting prod and the changes to the Onyx key.

**Review by mountiny — CHANGES_REQUESTED** [review]
In general, looks good to me, but there is one unit test, which was implemented for the currency list localization testing. 

https://github.com/Expensify/App/blob/c93b4cb8d3354addd5f2f8a289a80ea639a40a8a/tests/unit/CurrencySymbolUtilsTest.js#L4-L8

The instruction on how to obtain new JSON of the currencies uses our API so we should find a way how to keep this possible to be completed.

I am not sure how though as if we remove the API method from App then they can call it as they need authToken even if we keep the MobileConstants endpoint in the backend (or I mean they could use curl, and add their authToken but that would be quite cumbersome.

The best might be to update the instructions so they ask in the expensify-open-source channel so someone internal helps them retrieve the JSON. What do you think? Or just keep that change internal 🤷

**Comment by luacmartins** [comment]
Thanks for the review @mountiny! I updated the instructions. I think contributors can get the updated list from the OpenApp command right?

File: tests/unit/CurrencySymbolUtilsTest.js [hunk 2]
```diff
@@ -2,9 +2,10 @@ import _ from 'underscore';
 import * as CurrencySymbolUtils from '../../src/libs/CurrencySymbolUtils';
 
 // This file can get outdated. In that case, you can follow these steps to update it:
-// - in src/libs/API.js
-// - call: GetCurrencyList().then(data => console.log(data.currencyList));
-// - copy the json from console and format it to valid json using some external tool
+// - open your browser concole and navigate to the Network tab
```
Comment (mountiny) [inline tests/unit/CurrencySymbolUtilsTest.js:5]:
```suggestion
// - open your browser console and navigate to the Network tab
```

**Review by mountiny — COMMENTED** [review]
@luacmartins Good call, yeah they should!

Thanks for the changes, just one typo.

File: tests/unit/CurrencySymbolUtilsTest.js (hunk 2 above)
Comment (luacmartins) [reply tests/unit/CurrencySymbolUtilsTest.js]: 🤦 thanks!

**Review by mountiny — APPROVED** [review]
LGTM! Thanks for the changes!

**Comment by PauloGasparSv** [comment]
I'm still going to take a little time to review this to make sure this error is my fault and proceed. But right now It's giving me an error when I click to request money right after logging. If I refresh the error stops.
image

**Comment by PauloGasparSv** [comment]
Still have the same error! Can you help me on this @luacmartins?

https://user-images.githubusercontent.com/6564265/178829247-e81b0b5c-477f-459c-b73c-10a877423af2.mov

**Comment by luacmartins** [comment]
@PauloGasparSv the problem is that https://github.com/Expensify/App/pull/9560 is not merged yet. We set `selectedCurrencyCode` [here](https://github.com/Expensify/App/blob/10406bfdbdf1883cf27c74facec2d6b29ae541e1/src/pages/iou/IOUModal.js#L147), but it expects `localCurrencyCode` to be in the `myPersonalDetails` Onyx key. However, that key is now in `personalDetails`. Once that PR is merged it should fix this issue.

A side note is that [this default](https://github.com/Expensify/App/blob/10406bfdbdf1883cf27c74facec2d6b29ae541e1/src/pages/iou/steps/IOUAmountPage.js#L52) is doing nothing since other keys in `props.iou` are defined. We should probably fix that.

**Review by PauloGasparSv — APPROVED** [review]
I see! Nice : )
Approving it for now since everything LGTM!
But let's test it again when that P.R. is merged

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/pull/9560 merged! Retested and all is good. Removing hold and merging!

**Comment by luacmartins** [comment]
Tests passed! Removing `emergency` label!

**Pull request merged**